### PR TITLE
Enable OAuth

### PR DESCRIPTION
**Pull request by algomaster99**
### Testing the application

1. Backend
  i. Register a new OAuth application [here](https://github.com/settings/applications/new).
  ii. Write `CLIENT_ID` and `CLIENT_SECRET` in `.env` file.
  iii. A typical configuration can be this:
Screenshot from 2019-06-06 01-55-44

2. Frontend
  i) I created a button like this for testing:
auth
  ii) Write these two snippet in `/editor/src/components/header/renderer.tsx`
```
public getCookie(name) {
    const cookie = document.cookie.match('\\b' + name + '=([^;]*)\\b');
    return cookie ? cookie[1] : false;
  }
```
```
const auth = this.getCookie('user') ? (
      <form action="http://localhost:9000/auth/github/logout" method="get">
        <input type="submit" value="Logout" />
      </form>
    ) : (
      <form action="http://localhost:9000/auth/github" method="get">
        <input type="submit" value="Login" />
      </form>
    );
```
  iii) Reason for using forms for authentication instead of AJAX request is given in [this blog](https://medium.com/@wannabeph/xmlhttprequest-error-when-trying-to-google-login-a86620dbb0aa). Latter was creating CORS issues which was hindering the authentication process. Since header can't be modified when dealing with simple GET request using forms, it is considered more secure.
  iv) Login and logout to see the cookie set and unset.

I have also documented the code for better understanding.

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+   * @private
+   */
+  private initializeMiddleWares() {
+    // Configuration for creating session cookies
+    this.app.use(session({
+      key: 'user',
+      secret: sessionSecret,
+      resave: false,
+      saveUninitialized: true,
+      cookie: { httpOnly: false, maxAge: 1 * 24 * 60 * 60 * 1000 },
```
Comment (domoritz) [inline src/app.ts]: 1 hour? Please add a comment

File: tsconfig.json [hunk 2]
```diff
@@ -1,11 +1,18 @@
 {
   "compilerOptions": {
     "target": "es6",
-    "module": "es2015",
+    "module": "commonjs",
```
Comment (domoritz) [inline tsconfig.json]: I guess we need that for node?

Comment (algomaster99) [reply tsconfig.json]: Initially, it was throwing errors when I used `export`. This was the solution.

File: src/app.ts (hunk 1 above)
Comment (algomaster99) [reply src/app.ts]: @domoritz No, it is 1 day.

File: tsconfig.json (hunk 2 above)
Comment (domoritz) [reply tsconfig.json]: I think node doesn't understand es6 modules yet. deno does but I'm not sure we want to switch to it yet

Comment (algomaster99) [reply tsconfig.json]: I found [this link](http://wiki.commonjs.org/wiki/Modules/1.1.1) which explains how modules can be defined and used in `commonjs` format.

Comment (domoritz) [reply tsconfig.json]: Yeah, it's all gonna go away with es6 but that's okay. Keep this for now.

Comment (algomaster99) [reply tsconfig.json]: @domoritz `commonjs` only right?

Comment (algomaster99) [reply tsconfig.json]: @domoritz `import` in `es2015` isn't supported in Node. I think we will have to stick with commonjs. Reference: [https://stackoverflow.com/questions/39436322/node-js-syntaxerror-unexpected-token-import](https://stackoverflow.com/questions/39436322/node-js-syntaxerror-unexpected-token-import)

**Pull request merged**